### DISCUSSION TITLE
Named-only fields

### DISCUSSION THREAD

**Discussion by Ericson2314** [ghc-proposals/ghc-proposals#513] [Ideas]
Currently in Haskell today, you can name fields, but that doesn't prevent positional usage. Often in GHC MR review, I see folks asking "please change your positional exprs/pattrs to use field names", and they are right! It would be nice to allow enforcing this.

Besides arguably improving readability readability, a concrete benefit this allows is giving future versions the datatype definition the ability to freely reordering fields without risking breakage. Because that makes this a migration-oriented feature, I would include with https://discourse.haskell.org/t/pre-hftt-ongoing-focus-on-migration-tools/4626 if there is interest.

The solution I think would be a modifier on variants and (record-like) pattern synonyms to opt-in to this prohibition. Recently open https://github.com/ghc-proposals/ghc-proposals/pull/512 is also about data type modifiers, and so I would advocate making that and this share syntactic idioms.

**Comment by AntC2** [comment]
It's not that there would be anything wrong with a named-access-only record facility in Haskell -- if it had been built in at the beginning. Or if it had been added circa 2000 from the various records proposals then. But Haskell is not that language. If you want a language with decent handling of records, don't use Haskell; you'll just have to compromise on some other major features in another language.

There's been a flurry of proposals recently putting lipstick on the pig of Haskell's record-alike features. This one is more like putting lipstick on the deckchairs of the Titanic. Just stop.

**Reply by Ericson2314** [reply]
I believe ADTs with nice named fields nice and extensible records are two completely different features, and we're best off not conflating them. I want them both long term.

Yes, this is convenient in the short term, an incremental design refine. But per the above I think it's in service of a laudable not temporary stop-gap.

**Reply by googleson78** [reply]
I don't see how "the status quo is not great" implies "we shouldn't try to make small standalone incremental improvements". A big breaking change all at once seems very unlikely, while ditching Haskell for a language with type system parity+a better record system+library parity+ community and commercial adoption parity is a (long) while off.

**Reply by AntC2** [reply]
> [disagreeing that] "we shouldn't try to make small standalone incremental improvements"

Your metric is missing: "a large number of" small = low-benefit changes to the language I have to bear in mind. Each potential variation in allowed syntax is one more thing to pay attention to.

And one more option the compiler maintainers have to test in combination with all the other "small" (and not-so-small) options.

```haskell
{-# LANGUAGE  EnforceFieldLabels #-}

data Foo b = MkFoo { bar :: Int, baz :: b }

quux = MkFoo @Int 7 8                            -- Visible Type Application not allowed?

x = MkFoo $ z ^ v + w * u  $ "Charley"           -- to be rejected (presumably)
                                                 -- note MkFoo is an argument to $; $ is just some operator
justOK = Just $ z ^ v + w * u                    -- Just doesn't have labels declared, so mustn't reject this

mkFoo = MkFoo                                    -- allowed?
y = map (\ i -> mkFoo i "Farley") [1 .. 17]      -- reject passing the constructor to higher-order functions?

mkFoo2 :: forall b. Int -> b -> Foo b
mkFoo2 baa bah = MkFoo{ baz = baa, bar = bah }   -- what 'improvement' did we achieve?
doodle = mkFoo2 @Int 8 7                         -- now can use VTA

newtype FooStr = MkFooStr { unFooStr :: Foo String }  -- enforce labels for newtypes?
                                                 -- for datatypes with a single field?
                                                 -- for data families?
```

**Reply by Ericson2314** [reply]
> ```haskell
> mkFoo = MkFoo                                    -- allowed?
> ```
No it is not. Need the eta expansion
> ```haskell
> mkFoo2 baa bah = MkFoo{ baz = baa, bar = bah }   -- what 'improvement' did we achieve?
> ```
Upstream code and rearrange the fields and neither this function will its call sites will break. The tedious eta expansion is *locally* choosing a field order, which is good, because it is quite arguably *not* so good to have a bunch of random places like `MkFoo` being used as a higher order constructor relying on the global field order.

Fundamentally, this is trade off between brevity and fewer footguns when refactoring and reading code. Solo authors are likely to choose "brevity", but big projects and orgs are likely to choose "fewer footguns". I am not mandating anyone make a choice either way; I simply want to programming-in-the-very-large users better options.

> ```haskell
> quux = MkFoo @Int 7 8                            -- Visible Type Application not allowed?
> ```

This is interesting to think about. I think the solution is to generalize GADT record syntax some so we can do
```haskell
MkFoo { bar = 7, @baz_type = Int, baz = 8 }
```
I would say the fact that we didn't have this was already no good, and deserves its own issue, but I am glad this one somewhat "forces" the issue (though to be clear I don't think need or ought to block doing this on fixing that gap in expressive power.

**Reply by Ericson2314** [reply]
> ```haskell
> newtype FooStr = MkFooStr { unFooStr :: Foo String }  -- enforce labels for newtypes?
>                                                  -- for datatypes with a single field?
>                                                  -- for data families?
> ```
I said the annotation was opt-in and per data type. The user can choose to enforce or not enforce labels for all of these.

**Reply by googleson78** [reply]
> This is interesting to think about. I think the solution is to generalize GADT record syntax some so we can do

I think it makes sense to be able to say `MkFoo @Int <record syntax>`, because the `forall` is scoped on the constructor right now. 

(In a future where we have DH, this is simply solvable by adding a record adt containing `Type`s as arguments, made exactly as ergonomic as today's unnamed type args by potential anonymous records)

**Reply by Ericson2314** [reply]
`MkFoo @Int <record syntax>` is good too.

I had hesitated because I worried exposing anything positional was against the spirit of the proposal, but now I see since the allowed arguments are taken directly from type itself, and that *is* extremely exposed (`Foo Int` can't stop working), it is not such a stability issue.

**Reply by JakobBruenker** [reply]
> the allowed arguments are taken directly from type itself

This is not true for GADTs:

```haskell
ghci> data Foo a b where MkFoo :: forall b a . a -> b -> Foo a b
ghci> :t MkFoo @Int @String
MkFoo @Int @String :: String -> Int -> Foo String Int
```

**Reply by Ericson2314** [reply]
@JakobBruenker Yes, so I would only consider `MkFoo @Int <record syntax>` for Haskell-98-style data declarations.

**Reply by AntC2** [reply]
> I said the annotation was opt-in and per data type.

You said
>> a concrete benefit this allows is giving future versions of the code defining the datatype the ability to freely reordering fields without risking breakage. 

If some datatypes are declared with labels, some not, you can't freely reorder their fields. Then you still have to examine the whole module. You haven't made refactoring any easier.

If my datatype has some constructors with many fields, some with only one, I'm still forced to name the singletons(?) and use verbose syntax.

So as per @goldfirere's comment, you could teach the linter to look for data decls without named fields (meeting whatever extra criteria applies for house style).

**Reply by Ericson2314** [reply]
@AntC2 I mean the data type itself. I amended the description to make this clear. This isn't a per-module thing but a per-data or even per-variant thing.

> If my datatype has some constructors with many fields, some with only one, I'm still forced to name the singletons(?) and use verbose syntax.

Making it per variant helps.

**Comment by goldfirere** [comment]
For what it's worth, I'd be against this change. Seems quite possible to do in an external linter. Maybe if it's hard externally, we add a warning for this. But -- as much as I'd like to require named access in code bases I work in! -- I wouldn't want, say, a datatype designer to dictate this for all users.

**Reply by Ericson2314** [reply]
I in fact think linters could enforce it, but that doesn't help with the compatibility part of the motivation. For the upstream library to know it's safe to reorder fields, hard enforcement downstream code cannot opt out of is needed.

**Reply by adamgundry** [reply]
Perhaps I'm missing something here, because I don't understand this motivation. What's the point of the library author being able to reorder fields freely, if consumers can never use positional access anyway?

**Reply by AntC2** [reply]
He he. I guess it means a library author could add extra fields without consumers needing to change their code. Silently changing the semantics without telling consumers doesn't sound like a robust level of service. Will the datatype still work with `undefined`s in its fields?

**Reply by Ericson2314** [reply]
@adamgundry 

At first past the purpose of reordering the fields would just be cosmetic for haddocks and reader of the code, as new fields are added perhaps different fields make sense.

The larger story here, which we could make out of a few different proposals/issues, is making it is easier to change data types without breaking existing code in general, so fixing things like:

1. Cannot bundle patterns with type synoyms
2. No way to ask GHC whether it thinks my `{-# COMPLETE ... #-}` is correct
3. Patterns cannot help with record update

Writing these sorts of migrations is tricky no matter what, and so not having to worry about field order on top of everything else decreases the cognitive burden of doing so.

@AntC2 

> Will the datatype still work with `undefined`s in its fields?

We obviously don't want that! An explicitly bidirectional pattern synonym can pick defaults for the fields it doesn't expose.

**Comment by Ericson2314** [comment]
Here is a thought experiment I would like to ask those skeptical of this to consider: Suppose we had named-only fields today, and someone asked them to be positional too, how would that go over?

Every other recent language I know of has the field syntax mean order doesn't matter:

- Rust (example: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=4f8fcb5dedca33a8cd43e20bfe970a05)
- OCaml (https://v2.ocaml.org/manual/inlinerecords.html)
- JavaScript
- Python (it doesn't declaration side)

The only ones I know where it does are:

- C
- C++

This is also because "designated initializers" for structs were bolted on later.

My conclusion therefore is that exposing the positions of named fields is unexpected for users not familiar with the language, but sadly goes uncritiqued for existing users designing a named-field-literal system after the fact. I think accepting positions still matter is blind split of existing users who, used to positions already, don't realize how much this sticks out as weird from a fresh perspective.

**Reply by AntC2** [reply]
> Suppose we had named-only fields today, ...

But we don't today have anything remotely resembling Rows/Records/Variants.

> and someone asked them to be positional too, how would that go over?

I am not pleased about this deluge of records-oriented proposals where folks, say data designers, are trying to dictate a what-records-should-have-been for all users. We barely have named fields today. 

>  don't realize how much this sticks out as weird 

Everybody knows it's weird. You can see designs back to 1996 (probably earlier) trying to do something 'proper'. The H98 records design is a pig/it was a stopgap to go into the Language Report. You're trying to turn it into something it isn't -- by putting lipstick.

> exposing the positions of named fields is unexpected for users not familiar with the language,

There's lots of things unexpected for users not familiar with Haskell. Are you next going to introduce Objects? Or `GOTO`?

People who are not familiar with a language should learn to use it as it was 'designed' (using the word loosely). Or if they can't adapt, there's plenty other languages. This idea (and the pseudo-Variants proposal) just does not fit.

**Reply by AntC2** [reply]
Can I suggest a better way forward:

* Leave the H98 design as is, with all its warts.
* In fact start deprecating records/labels format, so ADTs are purely positional.
* We now have a mechanism for providing 'virtual'/views over datatypes -- that is, pattern synonyms.
* Propose a (new?) style of PatSyns that provides a bijective mapping to the warty ADTs.
* Then a module can export the PatSyns, but hide its ADTs/constructors.

**Reply by Ericson2314** [reply]
@AntC2 Focus on the Rust example then. Rust *also* don't have anything resembling extensible records / row typing. The Rust code I linked was an ADT variant with fields; one can have multiple ADT variants, they can be mixed positional / named fields just like us too!

I want extensible records too, but I also want what Rust has: non-extensible ADTs with field names and no positional usage. They are simply too different features to me. I am a little irked you seem to treat my desire for ADTs with named fields to be pure Stockholm Syndrome, and are trying to steer me towards extensible records + positional-only ADTs.

**Reply by AntC2** [reply]
I'm trying to steer you away from making an already messy bunch of compromises even more of a mess.

I'm also trying to point out that (contrary to what you seem to think) everybody knows it's a mess. It is what it is, put up with it. If anybody at GHC high command had been sufficiently aggravated about it, they'd have done something by now/there's been plenty of ideas/everybody knows enough about those other languages to mimic them if it was a dramatic urgency.

What I don't want is death by a thousand cuts/lots of itsy-bitsy options filling my head with low-benefit variations. (Like taking away selector functions from only some datatypes; insisting on named fields for other datatypes; other datatypes/constructors without named fields at all; others with fields named same as an unrelated datatype needing disambiguation sometimes; ...)

Haskell is already too hard to learn. In particular, too hard to get over the bump from newb level to intermediate. A rash of itsy-bitsy options makes it all the harder.

**Reply by Ericson2314** [reply]
> What I don't want is death by a thousand cuts/lots of itsy-bitsy options

If it weren't for back compat, I might flat out prohibit mixed named-position fields --- no new new options --- but I would *still* want this feature. I would also unconditionally turn on `-XNoFieldSelectors`. There, no new options.

This is Rust's ADTs work. 

To repeat myself, I want that. I want that in addition to extensible records. Not as a stop gap.

**Reply by AntC2** [reply]
> If it weren't for back compat, ... I would also unconditionally turn on `-XNoFieldSelectors`.

I'm with you there. I suggested that idea about 7 years ago.

I'd also disallow record update syntax.

IOW you can only use field labels:
* Inside a pattern match with explicit data constructor;
* Inside a record build with explicit data constructor.

Then we'd have no restrictions on duplicate record fields. Field puns and Record wildcards would work smoothly.

Clean. Simple.

Now that I've seen how the `-XOverloadedRecordDot` turned out, with its myriad of annoying limitations and the very unHaskelly `(.label)` accessor, I'd throw that away too.

>> It is what it is, put up with it.

I'd still want constructor-applied-positionally, which follows function-applied-positionally, including partial application.

**Reply by AntC2** [reply]
> what Rust has: non-extensible ADTs with field names and no positional usage.

I'm not getting what your gist is trying to show:

    let x = Bar::X(2,3);

looks exactly like positional usage -- of the fields inside constructor `X`, not using the declared field labels.

Neither do I get why your OCaml example is relevant: "constructors can now be defined using the same syntax as records" tells me OCaml does have stand-alone records (similar to purescript/Javascript).

    let v = Point {width = 10; x = 0.; y = 0.}

is constructor `Point` applied to a single argument, which happens to be a stand-alone record. "a unary constructor whose unique argument is a declared record type". (Presumably the implementation thereby has an extra data constructor for the record, as compared with Haskell constructors. [**])

      | Point p -> Point {p with x = l *. p.x; y = l *. p.y}

is similarly a case/pattern match with constructor `Point` bound to a single argument, which the usage on RHS shows to be a stand-alone record. I do like putting the record update gubbins [inside `{  }`](https://github.com/ghc-proposals/ghc-proposals/issues/328). Indeed I like it so much, I implemented it in Hugs a couple of years ago.

[**] Or perhaps the implementation is smarter than that: the `Error` example at the bottom is perhaps revealing the data constructor `Point` _is_ the record constructor.

**Reply by Ericson2314** [reply]
We agree with `-XNoFieldSelectors` and `-XOverloadedRecordDot`. Go hit "play" in the Rust example see witness the error.

**Reply by goldfirere** [reply]
I just want to chime in to say that I don't find the language "This comment exposes why your whole thinking is so wrong." particularly constructive. It seems the conversation became more productive afterwards (thank you), but it always dismays me to see ad-hominem language in these fora.

**Reply by AntC2** [reply]
Thank you for the guidance, Richard.

I've revised the comment along the lines ["I am not pleased about this deluge of issues where folks are trying to get ..."](https://github.com/haskell/core-libraries-committee/issues/73#issuecomment-1148233807); I've de-hominemed, to talk about [datatype designers](https://github.com/ghc-proposals/ghc-proposals/discussions/513#discussioncomment-2892594).
